### PR TITLE
Add new general and sequential executors.

### PR DESCRIPTION
**Pull request by xvallspl**
Disclaimer: not everyone tagged in this comment agreed to the totality of my changes, and all dissenting comments have been considered. Anything included this PR is carved in stone and everything is open for discussion

I'm proposing a generalized executor interface that resolves to the specific ones (```TThreadExecutor```, ```TProcessExecutor```, a new ```SequentialExecutor```) with an execution policy received as a parameter. This is a use case that we found in places such as ```TMVA``` (@omazapa, @lmoneta) and the various fitting functions, where currently we rely on several if-else instructions where we check the execution policy (or even if ROOT has been compiled with IMT) to instantiate the right executor.

My suggestion is to move ```TExecutor``` to ```ROOT::TExecutorBaseImpl``` (maybe in ```ROOT::Internals```?) and reuse the name. This breaks ROOT's very strict source compatibility requirements: "we should only break source compatibility if the ROOT constructs were actively harmful in some way, the volume of affected ROOT code is relatively small, and we can provide source compatibility and migration".

As discussed with @pcanal,  and with TExecutor introduced in ROOT 6.08, there's little possibilities any user has implemented a new derived class from TExecutor. If this happened, the user is most probably advanced enough to be able to change the implementation. In any case, the volume of affected ROOT code will be relatively small.  

As suggested by @dpiparo, for the moment we will keep the new ```TExecutor``` in ```ROOT::Internals```, not exposing it to the user. The executor usage will look as following:

```cpp   
ROOT::Internal::TExecutor pool(ROOT::Fit::ExecutionPolicy::kSerial);
auto mapFunction =  [](unsigned i){return 1u;};
auto reductionFunction =  [](const std::vector<unsigned> &v) {
      return std::accumulate(v.begin(), v.end(), 0u);
   };

pool.MapReduce(ROOT::Fit::ExecutionPolicy::kSerial, mapFunction, ROOT::TSeq<unsigned>(20), reductionFunction);
```

**TLDR;** This PR changes the behaviour of existing executors and introduces
new ones:

* TExecutor: changes its functionality to be a general Executor, while
TExecutorBaseImpl takes the role of the previous TExecutor. TExecutor
now acts as a general interface to the executors. The executor it will
resolve to is specified by a execution policy parameter in its
constructor.

* TExecutorBaseImpl: Plays the previous role of TExecutor.

* TSequentialExecutor: provides a sequential implementation of the
executor model, defined by TExecutorBaseImpl.

This PR is not finished, but I'm opening it for discussion. Things left:
- [x] Change kSerial to kSequential.

- [x]  Move Execution Policies to ROOT::Internal

- [x] Adapt Fitting functions to use the new TExecutor instead of if statements.

A test can be found here: https://github.com/root-project/roottest/pull/106

### CODE REVIEW

**Comment by dpiparo** [comment]
Hi, even before starting to comment this, for the sake of clarity, my proposal was not to go for the inheritance chain but rather for a utility function/class in ROOT::Internal which could absorb the duplication of code you described present in TMVA and fitting.

**Comment by xvallspl** [comment]
Hi, Danilo.

I missed the inheritance chain comment the first time, although I seriously considered your feedback. My justification for changing  the executors in 6.12 is that If we can reduce the damage to user code (e.g. reducing the time a soon to be changed interface is exposed and providing the new one) we should go for it.

But I'm not taking a hard stance on this and that's why I opened the PR, to discuss it :)

**Comment by xvallspl** [comment]
After talking this through, I agree this should wait until 6.14.

**Comment by pcanal** [comment]
Note: the failure in the TTabCom test indicates that as a consequence of this PR, more AST nodes are loaded in memory at startup.

**Comment by xvallspl** [comment]
I had this problem before when trying multiprocessing in the Fitting. I checked with Axel and, If I recall correctly, it was due to adding MultiProc as a dependency in MathCore.

I still kept the patch from then and I updated my roottest PR accordingly: root-project/roottest#106

File: math/mathcore/inc/Fit/FitExecutionPolicy.h [hunk 1]
```diff
@@ -1,9 +0,0 @@
-#ifndef ROOT_Fit_FitExecutionPolicy
-#define ROOT_Fit_FitExecutionPolicy
-namespace ROOT{
-   namespace Fit{
-      enum class ExecutionPolicy { kSerial, kMultithread, kMultiprocess };
-    }
-}
```
Comment (amadio) [inline math/mathcore/inc/Fit/FitExecutionPolicy.h]: Hi Xavi, I am not sure what @Axel-Naumann thinks about this, but I would try to avoid replicating in ROOT (with an incompatible interface) things that are part or will be part of the standard¹.

This should be a much easier backport than the troublesome `std::string_view`, and using the standard version has the advantage that ROOT algorithms can then forward the execution policy to std algorithms when that makes sense.

1. http://en.cppreference.com/w/cpp/algorithm/execution_policy_tag

Comment (xvallspl) [reply math/mathcore/inc/Fit/FitExecutionPolicy.h]: At this point I need to differentiate between multiprocessed and multithreaded (unsequenced) executions, and more that will come. There's certainly some ways we can do this besides creating our own execution policies, but each one I can think of has its flaws.

**Comment by lmoneta** [comment]
@dpiparo can you look at this one and decide if we need to close it ?

**Comment by xvallspl** [comment]
Will open a new version of this PR in the near future with the TExecutor discussed in PPP #67
